### PR TITLE
Add wait for K8s resources

### PR DESCRIPTION
**Pull request by schegi**
* Add some convenient methods for actively waiting for K8s resources cluster and nodepool the reach a specified state.
* Types K8S resource state constants

Bases on:  https://github.com/ionos-enterprise/ionos-enterprise-sdk-go/pull/80

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,15 +1,14 @@
...
-	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/go-resty/resty/v2 v2.3.0
 	github.com/jarcoal/httpmock v1.0.5
 	github.com/kr/text v0.2.0 // indirect
 	github.com/niemeyer/pretty v0.0.0-20200227124842-a10e7caefd8e // indirect
 	github.com/stretchr/testify v1.6.1
-	golang.org/x/net v0.0.0-20200625001655-4c5254603344 // indirect
 	gopkg.in/check.v1 v1.0.0-20200227125254-8fa46927fb4f // indirect
 	gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776 // indirect
+	k8s.io/apimachinery v0.18.6
```
Comment (avorima) [inline go.mod:11]: No, please don't import apimachinery here. I see you using one function from there, just re-implement it or write an equivalent function.

File: k8s.go [hunk 2]
```diff
@@ -3,6 +3,45 @@ package profitbricks
 import (
 	"fmt"
 	"net/http"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+type KubernetesClusterNodePoolState string
```
Comment (avorima) [inline k8s.go:11]: Remove this type

File: k8s.go [hunk 3]
```diff
@@ -3,6 +3,45 @@ package profitbricks
...
+	K8sStateFailedUpdating KubernetesClusterNodePoolState = "FAILED_UPDATING"
+	// Kubernetes cluster/nodepool resource state is destroying
+	K8sStateDestroying KubernetesClusterNodePoolState = "DESTROYING"
+	// Kubernetes cluster/nodepool resource state is failed_destroying
+	K8sStateFailedDestroying KubernetesClusterNodePoolState = "FAILED_DESTROYING"
+	// Kubernetes cluster/nodepool resource state is terminated
+	K8sStateTerminated KubernetesClusterNodePoolState = "TERMINATED"
+)
+
+type KubernetesNodeState string
```
Comment (avorima) [inline k8s.go:32]: Remove this type

File: k8s.go [hunk 4]
```diff
@@ -398,3 +437,33 @@ func (c *Client) ReplaceKubernetesNode(clusterID, nodePoolID, nodeID string) (*h
...
 }
+
+func (a *AutoScaling) Enabled() bool {
+	return !(a.MinNodeCount == nil || a.MaxNodeCount == nil || *a.MinNodeCount == 0 && *a.MaxNodeCount == 0)
+}
+
+func (c *Client) WaitForKubernetesNodePoolState(
+	clusterID, nodePoolID string,
+	state KubernetesClusterNodePoolState,
+	timeout, interval time.Duration) (err error) {
```
Comment (avorima) [inline k8s.go:448]: The error returned by this function should include which state was polled

File: k8s.go [hunk 5]
```diff
@@ -398,3 +437,33 @@ func (c *Client) ReplaceKubernetesNode(clusterID, nodePoolID, nodeID string) (*h
...
+	return wait.PollImmediate(interval, timeout, func() (done bool, err error) {
```
Comment (avorima) [inline k8s.go:449]: These named return values also are not needed

File: k8s.go [hunk 6]
```diff
@@ -398,3 +437,33 @@ func (c *Client) ReplaceKubernetesNode(clusterID, nodePoolID, nodeID string) (*h
...
+		var np *KubernetesNodePool
+		np, err = c.GetKubernetesNodePool(clusterID, nodePoolID)
+		if err != nil {
+			return false, err
+		}
+		return np != nil && np.Metadata != nil && np.Metadata.State == string(state), err
```
Comment (avorima) [inline k8s.go:455]: Return `nil` instead of `err` here

File: k8s.go [hunk 7]
```diff
@@ -398,3 +437,33 @@ func (c *Client) ReplaceKubernetesNode(clusterID, nodePoolID, nodeID string) (*h
...
+	})
+}
+
+func (c *Client) WaitForKubernetesClusterState(
+	clusterID string, state KubernetesClusterNodePoolState, timeout, interval time.Duration) (err error) {
```
Comment (avorima) [inline k8s.go:460]: Same remarks as above

**Review by avorima — CHANGES_REQUESTED** [review]
Needs rebase after https://github.com/ionos-enterprise/ionos-enterprise-sdk-go/pull/80 is merged

File: k8s.go [hunk 8]
```diff
@@ -444,26 +438,30 @@ func (a *AutoScaling) Enabled() bool {
...
+	if err := PollImmediate(interval, timeout, func() (bool, error) {
+		np, err := c.GetKubernetesNodePool(clusterID, nodePoolID)
 		if err != nil {
 			return false, err
 		}
-		return np != nil && np.Metadata != nil && np.Metadata.State == string(state), err
-	})
+		return np != nil && np.Metadata != nil && np.Metadata.State == state, err
+	}); err != nil {
+		return fmt.Errorf("error waiting for nodepool state %s [%w]", state, err)
```
Comment (avorima) [inline k8s.go:450]: Please use `: %w` for wraps

**Review by avorima — CHANGES_REQUESTED** [review]
Please run `go mod tidy` and `goimports`.

**Review by avorima — APPROVED** [review]
LGTM

**Review by pennycoders — APPROVED** [review]

**Pull request merged**